### PR TITLE
Introduce @LastElement annotation plus adapter

### PR DESCRIPTION
**Pull request by adauvalter**
Almost a pure copy of FirstElement but in my case I need a last element of a list :)

### CODE REVIEW

File: src/main/java/com/serjltt/moshi/adapters/LastElement.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+ *      .add(LastElement.ADAPTER_FACTORY)
+ *      .build();
+ * </code></pre>
+ */
+@Documented
+@JsonQualifier
+@Retention(RetentionPolicy.RUNTIME)
+@Target({ ElementType.FIELD, ElementType.METHOD })
+public @interface LastElement {
+  /** Builds an adapter that can process a types annotated with {@link LastElement}. */
```
Comment (serj-lotutovici) [inline src/main/java/com/serjltt/moshi/adapters/LastElement.java]: nit: types -> type/field

File: src/main/java/com/serjltt/moshi/adapters/LastElementJsonAdapter.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  private final JsonAdapter<List<T>> delegate;
+
+  LastElementJsonAdapter(Type type, Moshi moshi) {
+    Type listType = Types.newParameterizedType(List.class, type);
+    delegate = moshi.adapter(listType);
+  }
+
+  @Override public T fromJson(JsonReader reader) throws IOException {
+    List<T> fromJson = delegate.fromJson(reader);
+    if (fromJson != null && fromJson.size() > 1) return fromJson.get(fromJson.size() - 1);
```
Comment (serj-lotutovici) [inline src/main/java/com/serjltt/moshi/adapters/LastElementJsonAdapter.java]: `!fromJson.isEmpty()` is better, saves us a calculation.

File: src/main/java/com/serjltt/moshi/adapters/LastElementJsonAdapter.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    return null;
+  }
+
+  @Override public void toJson(JsonWriter writer, T value) throws IOException {
+    delegate.toJson(writer, Collections.singletonList(value));
+  }
+
+  @Override public String toString() {
+    return delegate + ".lastElement";
```
Comment (serj-lotutovici) [inline src/main/java/com/serjltt/moshi/adapters/LastElementJsonAdapter.java]: nit: missing parenthesis. Should be `".lastElement()"`.

File: src/main/java/com/serjltt/moshi/adapters/LastElement.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+import java.util.Set;
+
+/**
+ * Indicates that the annotated type/field is <strong>expected</strong> to be the last element of a
+ * json array.
+ *
+ * <p>For example if a json object is returned as:
+ * <pre>
+ *   [
+ *    {
```
Comment (vanniktech) [inline src/main/java/com/serjltt/moshi/adapters/LastElement.java]: Should not the example showcase an array of more than one element?

File: src/main/java/com/serjltt/moshi/adapters/LastElementJsonAdapter.java [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+import com.squareup.moshi.Moshi;
+import com.squareup.moshi.Types;
+import java.io.IOException;
+import java.lang.reflect.Type;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * {@linkplain JsonAdapter} that extracts the last element
+ * of an array of (a field) type annotated with {@linkplain LastElement}.
```
Comment (vanniktech) [inline src/main/java/com/serjltt/moshi/adapters/LastElementJsonAdapter.java]: The annotation can also be applied to a method

File: src/main/java/com/serjltt/moshi/adapters/LastElement.java (hunk 4 above)
Comment (adauvalter) [inline src/main/java/com/serjltt/moshi/adapters/LastElement.java]: Well, you can use if for the last element not only in array, but in object case also. It's totally the same like in firstElement.

Comment (serj-lotutovici) [inline src/main/java/com/serjltt/moshi/adapters/LastElement.java]: Adding this to the java doc wouldn't hurt.

**Pull request merged**